### PR TITLE
Boost: handle post edits and comments and delete cache files

### LINKED ISSUE #35055: [Boost] Delete cached content on edit.
Whenever the user performs an action which should invalidate cached content, delete the relevant cache files. Here is a starting point, though we should check Super Cache's codebase for additional cases where we should invalidate the cache:

- [ ] On page/post edit
- [ ] When you change your theme
- [ ] When you edit your page templates

Note that you shouldn't just invalidate the page being edited, it should also invalidate views which may show the relevant page.

### PR DESCRIPTION
**Pull request by donnchawp**
The Cache module has to delete cache files when posts are edited, comments are made or other actions are taken that modify the website.
This PR will take care of those actions and delete the right files so cache files don't grow stale.

Fixes [#35055](https://github.com/Automattic/jetpack/issues/35055)

## Proposed changes:
* Lots of error_log statements still in the code to aid testing. I'll have to remove them, and get an approval again.
* Hook onto various actions so the cache can be cleared when posts are edited, comments made or the theme is switched. We'll probably need more.
* In is_cacheable() check if the user is viewing a preview, or using the customizer. Not applicable right now because we don't cache for logged in users. Maybe in future?
* In normalize_request_uri(), parse_url may return null or false, so we need to handle those.
* Added the function is_visible_post_type( $post ) that returns true if the post type of that post is visible.
* Add the function delete_front_page_cache( $post ) because the front page should be handled separately. It can be in a number of formats (static page, or page of blog posts..)
* Added functions to handle comment editing and publishing.
* The post status can be either published or private which are visible on the web page, or a number of other statuses which aren't.  The delete_on_post_transition() function should take care of that and will delete the related tags and categories, and the homepage.
* delete_cache_for_post() was added to accept a $post, get a permalink, even if the post was trashed, and then delete the cache files for that post.
* delete_cache_for_post_terms() will delete the tag and category archives for the given post.
* delete_cache_for_url( $url ) deletes the cache files for that url, normalizing and then passing on the URI to the storage class.
* delete_cache() will delete the entire cache for the site.
* In the Utils library, the delete_directory() function gets a new parameter, $recurse. If it's 1 then it will delete any directories found by recursing down through them.
* delete_single_directory() simply calls delete_directory() with $recurse set to false.
* In the storage class, invalidate( $path ) will delete the path, recursively.
* invalidate_single_visitor() will delete the cache files for the given path, and user with GET and COOKIEs.
* invalidate_home_page() deletes the html files found in the root of the cache directory, and any "page" archives.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-2rA-p2

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:
* Create some tags.
* Create some test posts and add them to the tags.
* Populate the cache by visiting these posts, and the homepage too. You must use a private, NOT LOGGED IN, browser window.
* Tail the logs with `jetpack docker tail`.
* Edit the cache files and watch the tailed log. It should show it deleting that post, and the home page. Please verify the cache files were deleted from the directory.
* Change a post from published to draft. Change it from either of those to trashed. Restore it from trash. No cache files should be edited if the post isn't published either before or after.
* Leave a comment on a post. If moderated, you'll see a log entry saying the cache for that particular user was deleted.
* Try again.
* In wp-admin->comments, moderate the comments. Approve and reverse that. Edit the comment. Trash it. Examine the logs.
* Switch theme and note it deletes the entire cache.
* Link to another post on your site while editing a post. The moderated pingback will appear in the comments page. Approve it and note the cached pages being deleted.

### CODE REVIEW

**Comment by donnchawp** [comment]
Let's see what happens with https://github.com/Automattic/jetpack/pull/35515 before doing any more on this one.

**Review by dilirity — COMMENTED** [review]
Did a quick skim, haven't tested anything yet.

Do we need all the functions we hook on `init_actions` inside the `Boost_Cache` class?

We could attack this in an upcoming PR, but I was wondering if it makes more sense to have those things somewhere else, as they seem pretty specific.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Instead of `is_object( $post) `, I'd suggest using `is_a( $post, 'WP_Post' )`, as we're specifically looking for properties in there.

**Comment by donnchawp** [comment]
> Did a quick skim, haven't tested anything yet.
> 
> Do we need all the functions we hook on `init_actions` inside the `Boost_Cache` class?
> 
> We could attack this in an upcoming PR, but I was wondering if it makes more sense to have those things somewhere else, as they seem pretty specific.

@dilirity I was thinking about that too as it does make that file significantly longer. Maybe an "Edit" class? "Boost_Cache_Edit"?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Shouldn't deleting the front page always going to call `invalidate_home_page`? Maybe I'm missing something, but I'd think that's what it should do. Why do we need to clear the cache for a specific post?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: I think this is outdated, since `$post_id` is a `WP_Post` instance.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: I think naming can be improved here and they should all follow the same structure.

For example, `delete_cache_THING`.

Does this make sense?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php:386]: I get a headache just trying to understand this. It's not your fault, it's just these post transition hooks are such a messy thing.

The thing is now this, but is going to be that, but what was it before? 🤒

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: I've noticed that you haven't been adding proper docblock comments. They all should start with `/**`. That way, the editor properly picks it up and colors it, showing the variables and return statements in a neat color.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: As mentioned above, something to be mindful of. Add a second star here and see how the comment changes colors.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Lol, this function can basically return anything 😅

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php:309]: If a comment was `pending`, but was marked as `spam`, this would also reset the cache. Should it? I can't recall what happened when you added a comment to a page. I guess it shows up only for the user, but with a "pending approval" text.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php:407]: Hehe, nice thinking!

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php:368]: Why does this check for `private` as well?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: This is outdated. I don't see `$recurse`.

File: projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php]: We should make these translatable.

File: projects/plugins/boost/app/modules/cache/storage/File_Storage.php
Comment (dilirity) [inline projects/plugins/boost/app/modules/cache/storage/File_Storage.php:123]: This one should be colored in the editor!

**Review by dilirity — DISMISSED** [review]
Nice work on this! I haven't run any tests as I'm beat. I'll test it tomorrow.

I've left a few notes.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: If `post_id` is a `WP_Post` instance, should it be called `post`?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Is this different to `is_empty()`?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Belongs in `Boost_Cache_Utils`.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: I'm ok for the `error_log` statements to remain in here, pending being turned into proper logging statements as Adnan creates the logging system.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php:279]:
```suggestion
	 * Delete the cache for the front page if it will be affected by changes to the specified post.
	 * This is called when a post is edited, deleted, or published.
```

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: There is no `$all` param.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:407]: If we can end up invalidating the same post multiple times in a single PHP execution, could we also end up deleting multiple posts, multiple times: Is it worth storing an array of post_ids that have been invalidated this PHP execution to prevent re-deleting?

I don't know the answer to this question, btw - so if the answer is "no" that's ok.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:368]: `private` means it's published for administrators on the site. While we currently do not cache any logged in content, we might in the future.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php:367]: Should this go in `Boost_Cache_Utils`?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Nice; it's the complex cases like this I would never think of that I really appreciate your experience for :)

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/Boost_Cache.php:67]: Do we need `save_post` as well?

**Review by thingalon — CHANGES_REQUESTED** [review]
Oh, man. Working out what to delete for each kind of change is complicated in WordPress.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Good idea, done!

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: This is for when you're using a static page for the front page. (See Settings->Reading in wp-admin)

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Yes, definitely. I missed that when I updated that function.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Yep, renamed it to `delete_cache_for_front_page`.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:386]: The comparisons used to be a lot more complicated until I realised that I only needed to make sure that the post in question was never published, in the past, or if it was about to be.

In other words:
If a post was published and now isn't OR wasn't published and now is, then clear the cache for that post.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: oh, thanks. I never noticed that.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (dilirity) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Isn't the cache file the one for `/`?

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: I made the same error in Page_Cache and the Utils files but I won't fix them in this PR, until I need to modify them for something else that is and I remember seeing a comment about a function that should go in Utils ...

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Got it, and fixed that. Thanks!

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Yeah, basically empty(). I'll change it to that. It was annoying when I saw that parse_url returns so many false values.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:309]: You're right. I forgot about spam. There's also ['hold', 'approve', or 'trash'](https://developer.wordpress.org/reference/functions/wp_set_comment_status/). I'll have to add checks like in the post transition function.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:407]: I copied this from WPSC where there are more actions fired for deleting cache files, but since there are a few comment actions hooked, it's still needed.
I think it's ok to have just one. When I logged the deletes I didn't see the same post being deleted twice with this in place.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Thanks, fixed that.

File: projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache_Utils.php]: Good point. There are others that need translating too.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Good idea, done.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:279]: I decided to remove the code that checked if the post was on the front page because invalidate_home_page() now deletes the paged archives now, so the comment is correct.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Good catch, thanks!

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:367]: It probably doesn't because the only code that checks the status like this will be the post transition function.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: Thanks :)

File: projects/plugins/boost/app/modules/cache/storage/File_Storage.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/storage/File_Storage.php:123]: Fixed earlier.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:67]: I don't think we'll need that. The `transition_post_status` action fires when a post is saved, even if the [status hasn't changed](https://developer.wordpress.org/reference/hooks/transition_post_status/#comment-2977).

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:309]: Fixed that by checking if comment was not approved before and after, and returning if it was.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (thingalon) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: No; the site won't always be at `/`; it can be in a subdirectory, for example. 

But that then leaves me wondering how `invalidate_home_page` works, as it deletes the root of the cache directory, which may or may not be the site's homepage.

File: projects/plugins/boost/app/modules/cache/storage/File_Storage.php
Comment (thingalon) [inline projects/plugins/boost/app/modules/cache/storage/File_Storage.php]: If WordPress is installed in a subdir, or is otherwise not at the `/` URL of their site, then won't this completely fail to clear the cache of their homepage?

Also what is the check for `/page` for? I think that needs a comment explaining what it means.

**Review by thingalon — DISMISSED** [review]
Ok, I'm approving this because:
- I think the only issue remaining that I can see is that invalidate_home_page can't know that its home path is `/`
- I will be offline today while @dilirity and @donnchawp will be online to review and make sure this is good to go :) 

One note I will add: please be careful on how this merges with the hostnames update - https://github.com/Automattic/jetpack/pull/35647

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: You're both right. While I was sitting in the driver's seat in my car waiting for the roadside assistance yesterday, I tested using a static page for the front page. It was cached as "/", so that function can be simplified.
I'm going to test this PR on my test site in a subdirectory and see how that works and make adjustments.

File: projects/plugins/boost/app/modules/cache/storage/File_Storage.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/storage/File_Storage.php]: You're right about a subdirectory. I'm going to test that. /page/ is for paged archives. Here's [page 426](https://inphotos.org/page/426/) of my photoblog. :)

**Comment by donnchawp** [comment]
https://github.com/Automattic/jetpack/pull/35647 merged cleanly, thankfully.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: So. Two issues here:
* Site is hosted at /test/ instead of /
* Site is using a static page for the front page and another page for posts

I'll handle the second one first. The first "front_page_id" check above handled that, but it didn't work at all. It did update the static homepage, but never updated the page used to hold the posts.

I've just rewritten that. The static front page is at "/" (or "/test/", it doesn't matter as the post permalink has the right directory) and the posts page in my test example is at /posts/. Both locations are cleared when a post is updated. There might be a case to be made that the static homepage doesn't need to be updated, but some sites will want it updated.

Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php]: I just fixed having a site hosted at /test/ rather than / by passing the path from the home_url() to invalidate_home_page().

File: projects/plugins/boost/app/modules/cache/storage/File_Storage.php
Comment (donnchawp) [reply projects/plugins/boost/app/modules/cache/storage/File_Storage.php]: I just fixed that.

File: projects/plugins/boost/app/modules/cache/storage/Storage.php
Comment (pyronaur) [inline projects/plugins/boost/app/modules/cache/storage/Storage.php:14]: Because this is an interface, I would caution against requiring the Storage to know about specifics of what it's storing- "single visitor" or "home_page", that should probably be done at the implementation level.

Looking at the surrounding code, It looks to me like the `invalidate` interface should be `invalidate( $request_uri, $parameters)` and the rest should be handled by the implementation instead.

But I haven't followed along the developments here and I'm on SR right now, so just adding this as a comment here.

File: projects/plugins/boost/app/modules/cache/Boost_Cache.php
Comment (pyronaur) [reply projects/plugins/boost/app/modules/cache/Boost_Cache.php:407]: I don't know how this works, but it seems to me that doing this would break:

```
delete_cache_for_post( get_post(1) );
delete_cache_for_post( get_post(2) );
delete_cache_for_post( get_post(1) );
delete_cache_for_post( get_post(2) );
```

Even if the code works now, I would use an array of already deleted posts.

**Review by pyronaur — APPROVED** [review]
Approving this to allow merging, but I haven't reviewed this (I'm on SR) 🙈

File: projects/plugins/boost/app/modules/cache/storage/Storage.php
Comment (thingalon) [reply projects/plugins/boost/app/modules/cache/storage/Storage.php:14]: I understand where you're coming from, and I agree that we should think about what each class is responsible for.

But in this specific instance, I don't think you're correct. The Storage class is specific to caching, and exists to understand how to store data indexed by its metadata, in a format where it can have segments of its cache wiped based on their relationships.

I actually do think it's important for the Storage interface here to understand the semantics of the data it's storing, because the decisions it makes about how to store it directly stems from what the data is, and how different bits of the data relate to each other.

**Pull request merged**
